source/tests/system/visa_driver_api_tests.cpp

```diff
@@ -153,6 +153,149 @@ TEST_F(VisaDriverApiTest, SetNonTcpipAttribute_ReturnsAttributeError)
+  public:
+  VisaDriverLoopbackTest()
+  : portNumber_("1234")
+  , instrument_descriptor_("TCPIP0::localhost::" + portNumber_ + "::SOCKET")
+  {
+  }
+  virtual ~VisaDriverLoopbackTest()
+  {
+  }
+
+  void SetUp() override
+  {
+    start_server_session(portNumber_);
+    initialize_driver_session(instrument_descriptor_);
+  }
+
+  void TearDown() override
+  {
+    close_driver_session();
+    close_server_session();
+  }
+
+  void start_server_session(const std::string& portNumber)
+  {
+  #ifdef _WIN32
+    WSADATA wsa_data;
+    WSAStartup(MAKEWORD(2, 2), &wsa_data);
+  #endif
+    server_fd_ = socket(AF_INET, SOCK_STREAM, 0);
+    EXPECT_NE(-1, server_fd_);
+    struct sockaddr_in server_address;
+    server_address.sin_family = AF_INET;
+    server_address.sin_addr.s_addr = inet_addr("127.0.0.1");
+    server_address.sin_port = htons(std::stoi(portNumber));
+    EXPECT_EQ(0, bind(server_fd_, (struct sockaddr*)&server_address, sizeof(server_address)));
+
+    server_thread_ = std::thread(&VisaDriverLoopbackTest::run_server, this, server_fd_);
+  }
+
+  void run_server(SOCKET server_fd) {
+    listen(server_fd, 1);
+    SOCKET client_fd = accept(server_fd, NULL, NULL);
+    if (client_fd == -1) {
+      return;
+    }
+    std::thread([this, client_fd]() {
+      handle_client(client_fd);
+    }).detach();
+  }
+
+  void close_server_session()
+  {
+    server_thread_.join();
+  #ifdef _WIN32
+    closesocket(server_fd_);
+    WSACleanup();
+  #else
+    close(server_fd_);
+  #endif
+  }
+
+  void handle_client(SOCKET client_fd)
+  {
+    auto session = std::make_shared<TcpEchoSession>(client_fd);
+    session->start();
+  }
+
+  void write(const std::string& data)
+  {
+    auto response = client::write(GetStub(), GetNamedSession(), data);
+    EXPECT_EQ(VI_SUCCESS, response.status());
+    EXPECT_EQ(data.size(), response.return_count());
+  }
+
+  void read(uint32_t count, const std::string& expectedData, ViStatus expectedStatus)
+  {
+    auto response = client::read(GetStub(), GetNamedSession(), count);
+    EXPECT_EQ(expectedStatus, response.status());
+    EXPECT_EQ(expectedData.size(), response.return_count());
+    EXPECT_EQ(expectedData, response.buffer());
+  }
+
+  private:
+    std::string portNumber_;
+    std::string instrument_descriptor_;
+    SOCKET server_fd_;
+    std::thread server_thread_;
+};
+
+TEST_F(VisaDriverLoopbackTest, WriteAndRead_Matches)
+{
+  std::string writeData = "Visa gRPC read/write test";
+  write(writeData);
+  read(writeData.size(), writeData, VI_SUCCESS_MAX_CNT);
+}
+
+TEST_F(VisaDriverLoopbackTest, WriteTwice_ReadOnce_Matches)
+{
+  std::string writeData = "Visa gRPC read/write test";
+  write(writeData + "1 ");
+  write(writeData + "2");
+  std::string expectedReadData = writeData + "1 " + writeData + "2";
+  read(expectedReadData.size(), expectedReadData, VI_SUCCESS_MAX_CNT);
+}
+
+TEST_F(VisaDriverLoopbackTest, ReadMoreThanWritten_ReturnTimeout)
```

Comment:
Excellent test! I'm glad this works. I'd like to make the test method name a little clearer that it returns a response with what did get read along with the status. Maybe this?
`ReadMoreThanWritten_ReturnTimeoutErrorWithDataInResponsePacket`